Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,239 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic
+
+namespace Week2
+
+/-
+  # Logic: Implication, universal quantifier, existential quantifier, negation
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+/-!
+  # Implication and universal quantifier
+-/
+
+-- To write `∀` type
+
+-- Lean treats the universal quantifier as an implication.
+-- Implications are grouped on the RIGHT (N.B. this is different from the usual convention)
+
+theorem my_lemma : ∀ x y ε : ℝ, 0 < ε → ε ≤ 1 → |x| < ε → |y| < ε → |x * y| < ε :=
+  sorry
+
+-- Lean is all about types, and checking that something is of the right type
+section
+-- Naively, you can think of types as sets. Here these variables are real numbers
+variable (a b δ : ℝ)
+-- Being of the type of a Proposition means being a proof
+variable (h₀ : 0 < δ) (h₁ : δ ≤ 1)
+variable (ha : |a| < δ) (hb : |b| < δ)
+
+-- An implication is something like a function from sets of proofs to set of proofs
+#check my_lemma a b δ
+#check my_lemma a b δ h₀ h₁
+#check my_lemma a b δ h₀ h₁ ha hb
+
+end
+
+
+-- To prove an implication statement, we use intro
+
+example: ∀ x y z : ℝ, x ≤ y → y ≤ z → x ≤ z := by
+  intro x y z h₀ h₁
+  -- Now we can proceed as usual
+  apply le_trans
+  · apply h₀
+  · apply h₁
+
+
+-- Using the universal quantifier, we define an upper bounded function
+def FnUb (f : ℝ → ℝ) (a : ℝ) : Prop :=
+  ∀ x, f x ≤ a
+
+
+
+variable (f g : ℝ → ℝ) (a b : ℝ)
+
+-- It is useful to define functions, this can be done with the command `fun`
+
+#check fun x ↦ x + 1
+#check fun x ↦ f x + g x
+
+-- We can use implication hypothesis to get new hypotheses
+
+example (hfa : FnUb f a) (hgb : FnUb g b) : FnUb (fun x ↦ f x + g x) (a + b) := by
+  intro x
+  dsimp -- This step is needed for lean to unpack the definition of the function
+  apply add_le_add
+  exact hfa x
+  -- `hfa x` is now a proof of `f x ≤ a`
+  exact hgb x
+
+/-
+Since Lean sees an implication as a function from proofs of a Proposition to
+proofs of another Proposition, we can write the proof term explicitly, using `fun`.
+This is more concise, but probably less "human-like". Where to draw the line between
+conciseness and human readability is up to you!
+-/
+
+theorem fnUb_add {f g : ℝ → ℝ} {a b : ℝ} (hfa : FnUb f a) (hgb : FnUb g b) :
+    FnUb (fun x ↦ f x + g x) (a + b) :=
+  fun x ↦ add_le_add (hfa x) (hgb x) -- Note the brackets (terms would be grouped on the left by default)
+
+
+/-!
+# Exercises
+-/
+
+-- Let us define the property a lower bounded function
+def FnLb (f : ℝ → ℝ) (a : ℝ) : Prop :=
+  ∀ x, a ≤ f x
+
+
+-- 1.  Prove this using `intro`
+
+example (hfa : FnLb f a) (hgb : FnLb g b) : FnLb (fun x ↦ f x + g x) (a + b) :=
+  sorry
+
+-- 2. Now try proving it without using any tactic!
```

Comment:
what do you mean by without using any tactics?

Reply:
As a proof term

Reply:
Do you think it's confusingly worded or a confusing/too difficult task?

This is something it would be good if the students can do at some point (it shows they're really getting their heads around how lean understands lemmas etc), but open to the suggestion that we should cover it slightly later in the course.

Reply:
Happy for suggestions on the wording @themathqueen . For the question itself, is sort of advanced, but also there is an almost exact example in the demo, so I think they should be able to handle it

Reply:
Have suggested more explicit wording in another snippet.

Reply:
Yeah, I think it's the wording. Students might not _know_ what a "tactic" exactly is, yet.